### ISSUE TITLE
How to use driver-resolver with a JDBC connector?

### ISSUE THREAD

**Issue by krusaderjake** [tableau/connector-plugin-sdk#452]
**About You:**
Name: Jake Kruse
Industry: Transportation IT
I've recently taken an interest in creating a Tableau connector to SAS datasets using the sasiom (SAS Integrated Object Model) jdbc connection. I have a connector created with the relevant files that at least show the connector when I run my connector locally.

![image](https://user-images.githubusercontent.com/9046020/73112695-51d39b80-3ed5-11ea-8009-07a82319bfea.png)

I already have a detailed question on stackoverflow but I am not sure it will reach this specific community with such a specific question. https://stackoverflow.com/questions/59903764/tableau-driver-resolver-definition-for-jdbc-drivers-in-a-custom-connector

**Your question:**
**Overall, my question is, how can I confirm that my connector is using the correct driver and why a driver-resolver is not working to specifically pick the correct driver?** I want my connector to use the **com.sas.rio.MVADriver** but I see no reference to it in the logs. When I try to specify a driver-resolver in my .tdr file for exactly that driver, Tableau tells me to go download the drivers but I already placed the .jar files that work with the driver in C:\Program Files\Tableau\Drivers. Do the .jar files need to be placed elsewhere?

**sas.tdr file**
```xml
<tdr class='sas_jdbc'>
	<connection-resolver>
		<connection-builder>
			<script file="connectionBuilder.js" />
		</connection-builder>
		
		<connection-properties>
			<script file="connectionProperties.js" />
		</connection-properties>
	</connection-resolver>
	
	<driver-resolver>
		<driver-match>
		  <driver-name type='exact'>com.sas.rio.MVADriver</driver-name>
		</driver-match>
	</driver-resolver>
</tdr>
```

**Comment by pvanderknyff** [comment]
Hi Jake,

The driver resolver is an ODBC-only concept, the line in the API reference is a documentation bug. For a JDBC connector, there should be no driver resolver. I'm guessing that if you remove the driver-resolver from your current connector it should work, provided you set up your jdbc url correctly in your connection builder script: https://github.com/tableau/connector-plugin-sdk/blob/master/samples/plugins/postgres_jdbc/connectionBuilder.js#L2

**Comment by krusaderjake** [comment]
Okay, driver-resolver is ODBC only concept... I was following the docs here: https://tableau.github.io/connector-plugin-sdk/docs/api-reference

and they do state that driver-resolver is mainly used for ODBC but can be used for JDBC as well

![image](https://user-images.githubusercontent.com/9046020/73113498-e2ac7600-3ed9-11ea-8b9c-4f155a689688.png)

This is the error I get when I do not include driver-resolver which allows my connector to work but not work :)

![image](https://user-images.githubusercontent.com/9046020/73113525-096aac80-3eda-11ea-88ad-ea50ba0c39f5.png)

I've successfully been able to connect via JDBC to SAS using other software like SQuirrel SQL using the following JDBC connection string: **jdbc:sasiom://companyserver.company.com:port**

**In the tabprotosrv logs here is the well-formed JDBC string**

```
{"ts":"2020-01-24T18:45:07.163","pid":3980,"tid":"48c","sev":"info","req":"-","sess":"-","site":"-","user":"-","k":"msg","v":"JDBCProtocol::JDBCProtocol Connection URL: jdbc:sasiom://companyserver.company.com:port?","ctx":{"client-type":"desktop","procid":"16076","tid":"16496","version":"20193.19.1017.1259"}}
```

... of course I'm just excluding the actual server and port that jdbc:sasiom://companyserver.company.com:port represents

**Any idea what that http1.x error relates to? I thought it was because the correct driver is not being used but as long as the .jar files are listed in my environment in the Tableau logs then I should assume that Tableau is using the correct driver?**

**Comment by pvanderknyff** [comment]
I haven't seen that error before. Can you post your connectionBuilder.js script?

And thanks for pointing out the inaccuracy in the documentation, I've updated that part to reflect that driver resolver is an odbc-only component.

**Comment by jkoskela** [comment]
Can you provide clean logs? Delete all the logs first, and then try to connect, and attach those. Thanks.

**Comment by krusaderjake** [comment]
@pvanderknyff here is my connectionBuilder.js script:

```js
(function dsbuilder(attr) {
    var urlBuilder = "jdbc:sasiom://" + attr["server"] + ":" + attr["port"] + "?";
    var params = [];
    var formattedParams = [];
    for (var key in params) {
        formattedParams.push(connectionHelper.formatKeyValuePair(key, params[key]));
    }
	
    urlBuilder += formattedParams.join("&")
    return [urlBuilder];
})
```

if you need it here is my connectionProperties.js script:

```js
(function propertiesBuilder(attr) {
    //This script is only needed if you are using a JDBC driver.
    var params = {};

    // Set keys for properties needed for connecting using JDBC.
    var KEY_USER = "user";
    var KEY_PASSWORD = "password";

    // Set connection properties from existing attributes.
    params[KEY_USER] = attr[connectionHelper.attributeUsername];
    params[KEY_PASSWORD] = attr[connectionHelper.attributePassword];

    //Format the attributes as 'key=value'. By default some values are escaped or wrapped in curly braces to follow the JDBC standard, but you can also do it here if needed.
    var formattedParams = [];
    for (var key in params) {
        formattedParams.push(connectionHelper.formatKeyValuePair(key, params[key]));
    }

    //The result will look like (search for jdbc-connection-properties in tabprotosrv.log):
    //"jdbc-connection-properties","v":{"password":"********","s3_staging_dir":"s3://aws-athena-s3/","user":"admin"}
    return formattedParams;
})
```

@jkoskela - Attached are the log files

[tabprotosrv.txt](https://github.com/tableau/connector-plugin-sdk/files/4111570/tabprotosrv.txt)
[log.txt](https://github.com/tableau/connector-plugin-sdk/files/4111571/log.txt)
[jdbcserver.log](https://github.com/tableau/connector-plugin-sdk/files/4111572/jdbcserver.log)

**Comment by pvanderknyff** [comment]
Taking a look at the logs, thanks for providing them.

For the script, I don't know if this will change anything, but you can remove code for params and formattedParams and just return the urlBuilder variable. You can see an example [here](https://github.com/tableau/connector-plugin-sdk/blob/master/samples/plugins/postgres_jdbc/connectionBuilder.js).

Another thing you can try is to see if you can connect via [Other JDBC](https://help.tableau.com/current/pro/desktop/en-us/examples_otherdatabases_jdbc.htm) using your driver. This can show us this is a problem with the driver specifically or if the problem lies in the custom connector.

**Comment by jkoskela** [comment]
The error messages here indicate that there was a problem connecting to the JDBC driver sandbox (jdbcserver). It could be from a firewall. On Windows we have to connect using localhost TCP, because grpc doesn't support named pipes. Or it might be from classpath conflict. The classpath is pretty cluttered.

I would remove everything from Home/Tableau/Drivers, and just put on the jars needed for the driver in Program Files/Tableau/Drivers. Then unset the CLASSPATH environment variable so we don't pick anything up from there. Ideally the drivers would use a shaded fatjar so that it doesn't conflict with our stuff.

**Comment by krusaderjake** [comment]
Hi @pvanderknyff -

I tried removing the params and formattedParams from connectionBuilder.js and just returned the [urlBuilder] and still getting the same error.

Before I went down the custom connector route using the SDK I was getting the same "http1.x" error using the Other Databases (JDBC) connector. I posted about that here, on Tableau Communities: https://community.tableau.com/thread/328373

**Generically speaking with JDBC connectors (like the [Postgres_JDBC sample](https://github.com/tableau/connector-plugin-sdk/tree/master/samples/plugins/postgres_jdbc)) is the JDBC driver class name supposed to be referenced in the logs: com.sas.rio.MVADriver? I remember from testing Teradata ODBC and I actually saw the .dll Teradata driver that was used for connection.**

I'm still a little concerned that I don't see the SAS JDBC driver class name anywhere in the logs. Seems odd to me.

@jkoskela - Good notes and proxy/firewall could be part of the problem but I should be able to authenticate through. Are you saying it's more the jdbcserver that can't get through our firewall/proxy to make JDBC requests?

For now I know the .jar file clutter in the classpath that you're referring to. I'll clean that up, restart and run some more tests after I have a cleaner classpath. Please see my questions above.

**Comment by lriggs** [comment]
The driver manager can load the JDBC driver just using the URL if the driver is written correctly. It doesn't need to specify the class name. That is logged in the jdbcserver.log file after connecting. For example:
2020-01-28 16:44:10.811 -0800 (,,,,) grpc-default-executor-0 : DEBUG com.tableausoftware.jdbc.JDBCProtocolImpl - Trying class org.postgresql.Driver driver from: file:/C:/Program%20Files/Tableau/Drivers/postgresql-42.2.2.jar
2020-01-28 16:44:10.963 -0800 (,,,,) grpc-default-executor-0 : INFO  com.tableausoftware.jdbc.JDBCProtocolImpl - Connected using driver {org.postgresql.Driver}

You may want to specify debug level logging by starting Tableau with -DLogLevel=Debug

**Comment by krusaderjake** [comment]
Hi @lriggs - How does the driver manager load the JDBC driver only using the URL? I know that the driver is written correctly to make JDBC calls because I've been able to use the same driver class and .jar files with other applications outside of Tableau to connect to SAS and query for data.

I started tableau using the command line flag: `-DLogLevel=Debug` and I don't see the driver name listed in jdbcserver.log

**After using the debugging option here is what I see happening just walking through it:**
1. tabprotosrv log item: `Using JDBC driver directories` in the tabprotosrv file is finding the Drivers directory
2. tabprotosrv log item: `Enumerate jars in directory` is apparently looking in the Drivers directory at the .jar files
3. tabprotosrv log item: `JDBC Server starting` has the SAS .jar files listed
4. jdbcserver logs do not list the SAS driver

If the jdbcserver logs don't list the sasiom driver `com.sas.rio.MVADriver` then does that mean that the driver manager can't identify the JDBC Driver to use?

**Updated log files with debug option**
[tabprotosrv.txt](https://github.com/tableau/connector-plugin-sdk/files/4131569/tabprotosrv.txt)
[jdbcserver.log](https://github.com/tableau/connector-plugin-sdk/files/4131570/jdbcserver.log)

**Comment by lriggs** [comment]
Does your driver support JDBC 4+? Those are automatically registered. It should return the correct value in getJDBCMajorVersion and it looks like your driver needs to specify the class name internally: https://docs.oracle.com/javase/6/docs/api/java/sql/DriverManager.html

**Comment by jkoskela** [comment]
Those log messages like `com.tableausoftware.jdbc.JDBCProtocolImpl - Connected using driver {org.postgresql.Driver}` were added recently. They probably aren't available in the build you are using. The driver manager loads all classes that implement the Driver API, and then checks which one accepts the url scheme `jdbc:sasiom`. We don't use class.forName stuff anymore that is old JDBC style. There is no use for driver resolver for JDBC plugin. It will be removed from the docs.

**Comment by krusaderjake** [comment]
Hey @lriggs - I thought this Driver would be JDBC 4+ since it's the latest release as of Nov. 2018 but getJDBCMajorVersion returns 2, sadly. Looks like I'll have to wait for a release from SAS for a JDBC 4+ compatible Driver to accomplish this connection.

Thanks for all your help and ideas, but until I have a JDBC 4+ driver, it looks like I can't accomplish a SAS JDBC connection from Tableau.

Just out of curiosity, why does the Tableau SDK not allow for JDBC Type 2 Drivers and manual specification of the Driver class name? I imagine there are plenty of JDBC drivers out there that are below Type 4 and isn't Tableau meant for connecting to any data, wherever your data is?

**Comment by jkoskela** [comment]
JDBC v4 was released in 2006. This is the first v2 driver we have encountered.

**Comment by jkoskela** [comment]
Until SAS fixes their driver...you can embed a service manifest in the jar.

Add a file here:
/META-INF/services/java.sql.Driver

With contents:
com.sas.rio.MVADriver

**Comment by pschiltz** [comment]
@krusaderjake - did you surrender?  ever get this to work?  i  am trying to connect with the sharenet connection and am getting no where fast.
@jkoskela - there are about 17 jar files required for sas.  i placed them in a directory in the drivers directory.  which jar would i add the service manifest to?

**Comment by krusaderjake** [comment]
@pschiltz - Unfortunately yes, I did surrender. It wasn't in the cards for me to get this to work. I was using the sasiom jdbc connection that I could get to work in other applications but not here. I think I even tried what @jkoskela mentioned about embedding a service manifest in the .jar file but that hack did not work. I think it's just a result of JDBC versions type 2 (on SAS side) and type 4 (what Tableau expects) not playing together. Tableau expects the Driver Manager to register itself I believe and the hack did not work for me.

**Comment by krusaderjake** [comment]
@pschiltz - out of curiosity, have you gotten the sas sharenet connection to work in another application outside of Tableau?

**Comment by pschiltz** [comment]
Hi @krusaderjake -
Yes, I was so desperate to prove it COULD work that I used SAS/Access to JDBC to connect to the SAS JDBC driver to connect to SAS/Share!  So thats a lot of imbreeding.  But it worked when I finally put the right jars in the right place and got the right syntax on my libname statement.  I found reference internally that our driver (yes, I work at SAS) is type 4, so I just don't think that's it.

**Comment by krusaderjake** [comment]
@pschiltz - Here we are 8 months later and I decided to revisit a SAS IOM (sasiom) JDBC connector using the Tableau Connector SDK and I actually got it working to connect to the server.

1. I got past the http1.x error by making sure my proxy configuration was correct for our UNIX based SAS system that did not need to go through the proxy.
2. Then I was getting the error that a driver could not be found using the latest SAS JDBC drivers available here: https://support.sas.com/downloads/browse.htm?fil=&cat=50 so I still don't think that SAS is making JDBC Type 4.0 drivers available.
---> I had to embed the com.sas.rio.MVADriver text in the text driver file called java.sql.Driver in the sas.svc.connection.jar file under META-INF/services/java.sql.Driver file ... I created the folder called "services" and created the java.sql.Driver file with the driver name only in that file.
3. FINALLY, the DriverManager in Tableau was recognizing the sasiom jdbc subprotocol and driver and I could see it listed in the local Tableau jdbc server log file (which seems to have changed names since I was last looking at this stuff)
4. I'm able to connect but now I need to figure out how Tableau is interacting with SAS. I needed to create a connectionMetadata.xml file and include that in my Tableau connector manifest.xml so that when I connect to SAS I can see the tables available in the various "SAS libraries"
5. Now I'm dealing with various SQL translation issues where I can get a sheet open in Tableau but the Tableau SQL generated does not play nicely with how SAS interprets it I think via a live connection.

... Even though the ticket is closed, wanted to document this here and will post more when I figure out the SQL translation issues.
